exporters/otlp/otlp.go

```diff
@@ -0,0 +1,266 @@
+	errNotStarted     = errors.New("not started")
+)
+
+// Start dials to the collector, establishing a connection to it. It also
+// initiates the Config and Trace services by sending over the initial
+// messages that consist of the node identifier. Start invokes a background
+// connector that will reattempt connections to the collector periodically
+// if the connection dies.
+func (e *Exporter) Start() error {
+	var err = errAlreadyStarted
+	e.startOnce.Do(func() {
+		e.mu.Lock()
+		e.started = true
+		e.disconnectedCh = make(chan bool, 1)
+		e.stopCh = make(chan bool)
+		e.backgroundConnectionDoneCh = make(chan bool)
+		e.mu.Unlock()
+
+		// An optimistic first connection attempt to ensure that
+		// applications under heavy load can immediately process
+		// data. See https://github.com/census-ecosystem/opencensus-go-exporter-ocagent/pull/63
+		if err := e.connect(); err == nil {
+			e.setStateConnected()
+		} else {
+			e.setStateDisconnected(err)
+		}
+		go e.indefiniteBackgroundConnection()
+
+		err = nil
+	})
+
+	return err
+}
+
+func (e *Exporter) prepareCollectorAddress() string {
+	if e.c.collectorAddr != "" {
+		return e.c.collectorAddr
+	}
+	return fmt.Sprintf("%s:%d", DefaultCollectorHost, DefaultCollectorPort)
+}
+
+func (e *Exporter) enableConnections(cc *grpc.ClientConn) error {
+	e.mu.RLock()
+	started := e.started
+	e.mu.RUnlock()
+
+	if !started {
+		return errNotStarted
+	}
+
+	e.mu.Lock()
+	// If previous clientConn is same as the current then just return.
+	// This doesn't happen right now as this func is only called with new ClientConn.
+	// It is more about future-proofing.
+	if e.grpcClientConn == cc {
+		e.mu.Unlock()
+		return nil
+	}
+	// If the previous clientConn was non-nil, close it
+	if e.grpcClientConn != nil {
+		_ = e.grpcClientConn.Close()
+	}
+	e.grpcClientConn = cc
+	e.traceExporter = coltracepb.NewTraceServiceClient(cc)
+	e.mu.Unlock()
+
+	return nil
+}
+
+func (e *Exporter) dialToCollector() (*grpc.ClientConn, error) {
+	addr := e.prepareCollectorAddress()
+	var dialOpts []grpc.DialOption
+	if e.c.clientCredentials != nil {
+		dialOpts = append(dialOpts, grpc.WithTransportCredentials(e.c.clientCredentials))
+	} else if e.c.canDialInsecure {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
+	}
+	if e.c.compressor != "" {
+		dialOpts = append(dialOpts, grpc.WithDefaultCallOptions(grpc.UseCompressor(e.c.compressor)))
+	}
+	if len(e.c.grpcDialOptions) != 0 {
+		dialOpts = append(dialOpts, e.c.grpcDialOptions...)
+	}
+
+	ctx := context.Background()
+	if len(e.c.headers) > 0 {
+		ctx = metadata.NewOutgoingContext(ctx, metadata.New(e.c.headers))
+	}
+	return grpc.DialContext(ctx, addr, dialOpts...)
+}
+
+// Stop shuts down all the connections and resources
+// related to the exporter.
+// If the exporter is already stopped then this func does nothing.
+func (e *Exporter) Stop() error {
+	e.mu.RLock()
+	cc := e.grpcClientConn
+	started := e.started
+	stopped := e.stopped
```

Comment:
I think that the `stopped` member is not useful - it is used only in this function and it's changed from `false` to `true` only at the same time when `started` is changed from `true` to `false`. So the `if stopped` condition below is never evaluated to true. When `stopped` is `true` then `if !started` will be evaluated first and function will return an error.

This also means that either docs should say that the function returns an error if the exporter is stopped (so it matches the code), or the `if !started` condition should do `return nil` to match the docs.